### PR TITLE
Use new style dialog to browse a folder

### LINKED ISSUE #8513: Please extend the "Browse for Folder" to match path opened
## Issue:
The "Browse for Folder" is the lazy browser instead of the full one -- I can't drop a path, so as coders that likely have a deep hierarchy, we have to manually find what we're looking for.

Let's say I have a path, but forgot the file name. Paste the path in >> browse >> it goes straight back to the beginning "Desktop, etc" O_o

## Solution
Either use a full file browser popup, or allow browse to start from the path input instead of top level.

### PR DESCRIPTION
**Pull request by mere-human**
Add new methods to CustomFileDialog:
- setTitle() sets the dialog title
- setFolder() sets an initial directory
- pickFolder() shows a file open dialog to select a folder

Use CustomFileDialog in folderBrowser().

Affected areas:
- Search > Find in Files
- File > Open Folder as Workspace
- Preferences > Default Directory
- Preferences > Backup > Custom Backup
- Preferences > Cloud & Link

Fix #8513

### CODE REVIEW

File: PowerEditor/src/MISC/Common/Common.cpp
Comment (mere-human) [inline PowerEditor/src/MISC/Common/Common.cpp:178]: Do we really need to hide the "New folder" button?

File: PowerEditor/src/MISC/Common/Common.cpp
Comment (mere-human) [inline PowerEditor/src/MISC/Common/Common.cpp:155]: Do we need to convert `getFolderName()` to use the same CustomFileDialog?
That function is defined after this one.

**Comment by mere-human** [comment]
New Browse for File dialog looks like this:
image

File: PowerEditor/src/MISC/Common/Common.cpp
Comment (sasumner) [reply PowerEditor/src/MISC/Common/Common.cpp:178]: NO!

**Comment by sasumner** [comment]
>New Browse for File dialog looks like this

Will it accept paths (or partial paths) with `/` as directory separator, in the `Folder:` box?  :-)

**Comment by sasumner** [comment]
> Affected areas:

With all of these areas, it is getting harder to hold on to the existing regular File-Open and File-SaveAs dialogs (you know, the workhorse ones).  :-)

**Comment by sasumner** [comment]
BTW, I would have done all of these types of dialog changes myself some time ago, but @donho 's dislike of COM (aka the "dark power" LOL) prevented me from putting the effort in.  :-)

**Comment by mere-human** [comment]
> Will it accept paths (or partial paths) with `/` as directory separator, in the `Folder:` box? :-)

No, not yet. I plan to solve this issue separately, in scope of #9374

File: PowerEditor/src/MISC/Common/Common.cpp
Comment (donho) [reply PowerEditor/src/MISC/Common/Common.cpp:155]: Yes, please! 👍 
Or better : just replace it by `folderBrowser()` if it is it possible.

**Review by donho — CHANGES_REQUESTED** [review]
Replace or convert getFolderName() with new folderBrowser()

**Comment by donho** [comment]
@sasumner 
> BTW, I would have done all of these types of dialog changes myself some time ago, but @donho 's dislike of COM (aka the "dark power" LOL) prevented me from putting the effort in. :-)

I didn't recall that you wanna give it a try. Sorry about that. 
But since @mere-human came with an entire solution, and it seems not so "dark" to me, it's easier to accept the solution - especially there's no other remedy for that.

File: PowerEditor/src/MISC/Common/Common.cpp
Comment (mere-human) [reply PowerEditor/src/MISC/Common/Common.cpp:144]: Now `getFolderName()` just calls `folderBrowser()`. It behaves like a shortcut with a predefined parameters like the title.

**Comment by mere-human** [comment]
> Replace or convert getFolderName() with new folderBrowser()

Done. It was used only in Project Panel to add files from a directory.

**Comment by sasumner** [comment]
> I didn't recall that you wanna give it a try. Sorry about that.

Ha!  Well, you don't read minds do you?
It was all in my head about doing such a thing.  I never said anything about it.
Thus nothing to be sorry about.

**Review by donho — APPROVED** [review]

**Comment by mere-human** [comment]
One thing that bothers me a bit is that the original folder picker functions didn't respect the "Default Directory" preference.
In some cases, the initial directory was provided as an argument (_Find in Files_) but not always.
For example, if you open _File > Open Folder as Workspace_ it will always start from the top-level directories (This PC, User, etc).
However, for some users, this might be expected behavior.
So I decided to preserve the same behavior for the new folder picker.

**Comment by donho** [comment]
@mere-human 
I don't quite follow you - could you provide me a concrete example?

**Comment by mere-human** [comment]
> @mere-human
> I don't quite follow you - could you provide me a concrete example?

I mean, when we set the _Default Directory_ preference, then most of the file dialogs will set that as an initial directory.
image
For example, _File > Open_ will show that directory:
image
However, this is not the case for folder picker dialogs. An example is _Open Folder as Workspace_:
image
This was the behavior of the old dialogs and I preserved the same in this PR.

But now that I think about it, it seems to be a different, not related problem.

**Comment by donho** [comment]
OK, I see.
The Default Directory preference is only for open or save file dialog.
It's not for applying for the folderBrowser IMO.

**Comment by donho** [comment]
@mere-human 
FYI (to not have conflict for you):
https://github.com/notepad-plus-plus/notepad-plus-plus/commit/e26199ab51258fa6f743566aa6018eed8fb64b4f

**Comment by Uhf7** [comment]
I just discovered that using the new style dialogs makes Notepad++ unusable under Windows XP, because these functions are only available since Windows Vista. Under Windows XP, Notepad++ dies now with a message box saying "The procedure entry point SHCreateItemFromParsingName could not be located in the dynamic link library SHELL32.dll".

**Comment by sasumner** [comment]
>using the new style dialogs makes Notepad++ unusable under Windows XP

XP support was dropped a bit ago; @chcg recently told us exactly with which commit it was dropped but I can't find it ATM.

**Comment by mere-human** [comment]
@Uhf7 @donho Any idea how many Windows XP users does Notepad++ have?
Although, Windows XP is not officially supported anymore.
If this is a concern, we can revert my changes for new style dialogs. Or I can implement a workaround specifically for Windows XP users.

**Comment by sasumner** [comment]
>we can revert my changes for new style dialogs.

IMO that doesn't make sense

>Or I can implement a workaround specifically for Windows XP users.

IMO: Could be done, but it sounds like complicating the code unnecessarily.

IMO: XP users want to use an ancient OS that is no longer updated by its maker, they can use older versions of N++ (also which are not updated by _ITS_ maker) that DO support their OS.

**Comment by Uhf7** [comment]
I don't use Windows XP anymore, but obviously some users do. I stumbled over it in Issue #9347.

I brought up the question, because Notepad++ 7.9.2 still works under Windows XP, and the next version will not start anymore. This is the hard end of XP support. If everyone is fine with this, I don't need XP neither.

**Comment by sasumner** [comment]
> 7.9.2 still works under Windows XP

When I said earlier that "@chcg recently told us exactly with which commit it was dropped" I meant that there was a feature that was no longer working when invoked, which is obviously a bit different than "won't start anymore".

**Comment by Uhf7** [comment]
Perhaps it deserves a bigger jump in the version number, like to 8.0.0. And this "completely new" version doesn't run under Windows XP anymore. And the preprocessor definition WIN32_WINNT=0x501 could be upgraded to WIN32_WINNT=0x600. Then, the redefinition of WIN32_WINNT here
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/41c4180b2ef75e963645baca2fcfa0e853f64365/PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp#L31

would become unnecessary, and we would have a software designed to run from Windows Vista on.

It's an administrative decision. I imagine, many XP users didn't realize the end of support until now, but they will realize it, as soon as it doesn't start anymore.

**Comment by sasumner** [comment]
FWIW [wikipedia](https://en.wikipedia.org/wiki/Windows_XP) says "As of October 2020, 0.8% of Windows PCs run Windows XP" -- do we hold back nice new features from the other 99.2% of users just for the 0.8%? :-)

**Comment by donho** [comment]
I would say, if Notepad++ still runs under XP, it'll be shame to stop running under this OS, because only this new folder dialog.
OTOH, we can stop "supporting" XP (for example the bug **Focus does not return to a Project panel #9347** under XP won't be fixed.

We can still keep the new folder dialog, all we need to do is switch to old folder dialog if it's under XP. The [detection of XP](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/NppCommands.cpp#L3062) is quite simple, and we still get the old folder dialog code - so why shouldn't we?

**Comment by donho** [comment]
> Under Windows XP, Notepad++ dies now with a message box saying "The procedure entry point SHCreateItemFromParsingName could not be located in the dynamic link library SHELL32.dll".

OTOH, I have a doubt - once Notepad++ contains the new folder dialog code, the above message will display, even we have `if` for not calling it? From the description provided from @Uhf7 , it seems SHCreateItemFromParsingName should be located from the beginning.

**Comment by Uhf7** [comment]
There is the preprocessor variable WIN32_WINNT. It determines, which system functions can be used and which not. WIN32_WINNT=0x501 means, only XP functions can be used. If functions of newer systems are used, compiler errors occur. That's why this preprocessor definition has been redefined in `CustomFileDialog.cpp`.

The optional usage of the new file dialog functions may be possible, although complex. The missing parts in XP are `IFileDialog` itself and it's descendants, and some system functions which could be loaded dynamically with `LoadLibrary`. 

One main problem seems to be testing and verifying that Notepad++ can really start under XP. As long as `WIN32_WINNT` is never set to a value greater than `0x501`, it is quite sure, that the software will start under XP. As soon as this is set to a greater value anywhere in the code, the resulting exe file could contain static links to system library functions which doesn't exist in XP.

But `IFileDialog` cannot be used with `WIN32_WINNT` set to a value lesser than `0x600`. To create an XP-save version, own header files would have to be created to access this, on the base of the GUID of the `IFileDialog` interface. And the definition of the data structures needed would have to be cloned too.

If this is done, it can be checked at run time, whether the `IFileDialog` interface does exist, and it can be used on newer-than-XP systems only. Perhaps, there is an easier way, but I don't see it.

**Comment by sasumner** [comment]
>so why shouldn't we?

It all sounds very complicated, especially to maintain.
Because I've had PRs rejected because the code was too complicated to make it worth it, I'm very sensitive to this.  :-)
I have no more opinion on this.

**Comment by donho** [comment]
I didn't find any way to use `IFileDialogCustomize` with _WIN32_WINNT=0x0051 other than what @Uhf7 has suggest. So I guess that we have to drop XP :(

> Perhaps it deserves a bigger jump in the version number, like to 8.0.0. And this "completely new" version doesn't run under Windows XP anymore. 

Dropping an old OS is not like starting to support a new OS, so IMO it's not necessary to jump in a major version.
OTOH, user should be informed via release news & change log.

> And the preprocessor definition WIN32_WINNT=0x501 could be upgraded to WIN32_WINNT=0x600. Then, the redefinition of WIN32_WINNT here

I'll ask @mere-human to remove them in his future PR.

**Comment by mere-human** [comment]
I created a PR #9412 in which I address the Windows XP startup issue.
Please, have a look
However, the new style dialogs won't open on Windows XP but that can be addressed as well.
At least, the application is able to launch.

**Comment by fml2** [comment]
> Dropping an old OS is not like starting to support a new OS, so IMO it's not necessary to jump in a major version.

It would be a "breaking change" IMO (something that worked before will not work anymore) which requires a new major version, if one uses something like semantical versioning. But NP++ does not use it, right?

**Comment by Hermholtz** [comment]
I think abandoning XP warrants changing the major version of Notepad++ to 8.0.0.

**Comment by fml2** [comment]
Yes. See 7.9.3 :-)

**Comment by sasumner** [comment]
@Hermholtz Apparently Don disagreed.

**Comment by Hermholtz** [comment]
Yeah, so I've looked what was the main selling point for changing major version to v7. And it - lo and behold - 64bit build.

**Comment by Terwilf** [comment]
Hello, I am one of the few who use XP, and the conclusion of this thread seems very reasonable to me, so I did not want to miss the opportunity to thank so many years of support, thank you very much and best regards, from Argentina

**Comment by zahra21** [comment]
hello.
i am one of the lovers of windows xp.
xp is the most simple and user friendly operating system in the world.
i became very sad when i reads what's new of notepad plus plus 7.9.3
i am not professional notepad user.
i use notepad for reading and writing files only.
i read the comment on this page and related issues.
but i did not understand exactly what happened!
did you remove support for windows xp entirely? and notepad cant be installed and used on xp even for simple tasks?
or, just new method for project panel which is available on vista and newer, does not work on xp?
i never used project panel and dont know what is it!
i wish that notepad supports xp as its supports in version 7.9.2 at least the features that they are not dependant on windows version!
(the features which work in all versions of windows).

**Comment by zahra21** [comment]
by the way, scintilla still supports windows xp.
may i ask what is the version of scintilla in notepad 7.9.2?

**Comment by sasumner** [comment]
@zahra21 

>i did not understand exactly what happened!

A code change was introduced to use features that aren't present in Windows XP.
The result is that Notepad++ 7.9.3 (and presumably later) won't even start up on XP.

>what is the version of scintilla in notepad 7.9.2

4.2 , which can be seen here:

image

**Comment by zahra21** [comment]
since i am blind, can you please describe the image for me?
how can i know about the new code in vista and later versions
and its difference between xp?

**Comment by sasumner** [comment]
>can you please describe the image for me?

The image is what appears when, using Windows Explorer, you right-click on Notepad++'s SciLexer.dll file, choose Properties, then select the Details tab.
Mainly I was showing how you'd see the Scintilla version, which, as I said, is 4.2 (in Notepad++ 7.9.3).

>how can i know about the new code in vista and later versions and its difference between xp?

By researching such differences on the internet, I suppose.

**Comment by zahra21** [comment]
i ment scintilla version number in notepad 7.9.2
is it 4.2 or other previous versions?

**Comment by sasumner** [comment]
>i ment scintilla version number in notepad 7.9.2
is it 4.2 or other previous versions?

Here are my notes on this (not guaranteed to be 100% accurate):

```
Scintilla version bumps:
N++ 7.8 - Scintilla 4.2.0
N++ 7.7 - Scintilla 4.1.4
N++ 7.6.6 - last to use Scintilla 3.5.6
```

As this discussion is off-topic, if you have more questions, please take them to the Community site : https://community.notepad-plus-plus.org/

**Comment by Deadooshka** [comment]
Stand with Windows XP

**Comment by HBelusca** [comment]
Otherwise using delayloading for these functions... nobody thought about that?

**Comment by mere-human** [comment]
> Stand with Windows XP

You can continue using Notepad++ version 7.9.1 or 7.9.2 on Windows XP. It works perfectly fine.

> Otherwise using delayloading for these functions... nobody thought about that?

There was such an idea: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9412/
But it was decided not to overcomplicate matters. 
If not this feature, something different would prevent Notepad++ from running on Windows XP.
Supporting the application by a few contributors in their free time and such many bugs/features is a burden already.

I wonder, how many applications still support Windows XP? Browsers, for example. 
Does something different from IE work on XP?

**Comment by Terwilf** [comment]
> Does something different from IE work on XP?
Maxthon, 360Extreme (360极速浏览器) y Serpent, among others...

I reiterate, I do not see the need to continue receiving updates, but we must not trivialize things, teams with low resources are a reality in underdeveloped countries.

On the other hand, Microsoft revealed that a total of 1500 million users make use of computers based on any of the versions of its operating system, taking into account that the figure used by Wikipedia belongs only to the world market share of the version of Windows desktop (Source: Statcounter)

We could say that 0.8% corresponds to a total of 12,000,000 people, not to mention that logically only computers with an Internet connection can be counted...

**Comment by korick** [comment]
i love xp.
Don't drop support in minor version 7.9.3.
Drop support in major "completely new" version like 8.0.0.
Why not ?
[url=https://github.com/Terwilf]Terwilf,[/url] Yes you are right.
off-top

**Comment by zahra21** [comment]
@sasumner,
yes, your informations was correct.
you said exactly the informations which i needed.
about my other questions, can i send email for you or other developers?
i reported an accessibility issue in notepad++ and did not recieve any comment!thanks so much for your help and God bless you!

**Comment by sasumner** [comment]
>about my other questions, can i send email for you or other developers?

If you truly have only "questions", it is best to take them to the Community site:  https://community.notepad-plus-plus.org/

If what you are talking about is a level above "questions", i.e., "issues", it is most organizationally efficient to keep addressing them here in github.  Even if you think something is above the level of a question, sometimes talking it through with Community site members can help refine it.

Dropping to email is a recipe for lack of tracking and things getting lost.

>i reported an accessibility issue in notepad++ and did not recieve any comment!

We have a lot of issues and a limited staff, so unfortunately this kind of things happen.
You can always "freshen" an issue by posting an "any updates?" to it, but I wouldn't do that too often.

**Comment by zahra21** [comment]
i have three issues when finding my desired words in notepad++ with using nvda screen reader and i reported one of them on github several days ago.
the last isssue i think is resolved in 7.9.2 but i am not completely sure.

**Comment by sasumner** [comment]
@zahra21 Well it doesn't do any good to discuss it in an unrelated issue.

**Comment by zahra21** [comment]
mere-human,
"Stand with Windows XP
You can continue using Notepad++ version 7.9.1 or 7.9.2 on Windows XP. It works perfectly fine."
thanks so much for your great program, i really love notepad++.
i also wish that newer versions continue to support xp too.
"There was such an idea: #9412
But it was decided not to overcomplicate matters.
If not this feature, something different would prevent Notepad++ from running on Windows XP."
i really wish that you implement fix for issue 9412 that notepad can run again on windows xp perfectly fine.
"Supporting the application by a few contributors in their free time and such many bugs/features is a burden already."
on the contrary, i believe that notepad++ is a very stable program, i never encountered any crash or slowness when using notepad.
its one of the most stable, has less bugs and its very trustworthy.
i saved hundreds of files with it without any issues.
I wonder, how many applications still support Windows XP? Browsers, for example.
i found some file manager programs that even support windows 95 and 98!
but i always thought that because notepad++ is very stable and bugs are few in notepad, it continues to supports windows xp forever!
i sincerely pray for you and all developers of my favorite programs like notepad++ and nvda screen reader.
God bless you all and His infinite mercy i pray for you!

**Comment by KOLANICH** [comment]
I think it is complete disrespect to readers of the changelog to write that XP support was dropped because of "technical reasons". It is lie: the reasons are not technical, but IT-consumerist ones:

* The essentional functionality of a text editor is ... a working text editor.
* Usage of a new-style file dialog is not an essential functionality. It is just bells-and-whistles that are nice to have.
* Sacrificing essentional functionality to get bells-and-whistles is strange.
* Especially taking in account that `LoadLibrary` is still here and can be used to import the API dynamically, and rollback to other API, if the newer one is not available.
* >>using the new style dialogs makes Notepad++ unusable under Windows XP
  >Good!
  Nice point: the app got an antifeature making it completely unusable for some users and this is called "good".
* https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9412#issuecomment-763296464
* https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9412#issuecomment-762759603
* All the previous points mean that the decision was political, not technical. As the ones maintaining the project and actually writing the code and bearing expenses, you have a full right to drop all the systems, architectures, locales and CPUs you like, you also have full right to integrate malware in your software, GPL allows you to do it and be spared of any liability: the software is distributed AS IS without any warranty and liability. But please, **don't pretend that readers of the changelog are idiots. Consumerism of the devs and maintauners of this repo is not a technical reason. Please write clearly something like: "we, devs of this repo, have decided, that we don't want support the software and hardware we personally don't need. We understand that some users will be either forced to buy new software and hardware or maintain and build a fork doing that. We just don't care. We are (not) ready to keep supporting that hardware and software, (even) for $N per month."**

**Comment by Hermholtz** [comment]
I think the best solution here to continue supporting XP would be to offer a pull request. If anyone feels qualified, please do so. The author may be overworked, if you can help them then please do.

**Comment by KOLANICH** [comment]
There already [was a PR](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9412), though I personally haven't tested it personally  (TBH I have never build NPP myself) and IDK if it really works.

**Comment by Terwilf** [comment]
I think they are exaggerating a bit, since if the update is merely cosmetic, it is because this program has already covered practically all the possible functionalities that a text editor can have, I do not understand why instead of thanking that this software maintained the support for a system that was decommissioned 7 years ago, they reproach it for wanting to continue growing.

In addition, the source code is public, that is, its way of proceeding is completely transparent, I mean if it were really a popular demand, practically anyone could create their own compilation of the software.

I say this as a Windows XP user, since I have no other option, that is why I know and see in first person how limiting it is to develop for this system, when my brother helps me develop applications, sometimes he has to do several compilations until he achieves that everything works fine on my OS, it's frustrating, but it is.

python, .net framework and many lesser known libraries no longer support XP, it is sad as I said, for many this operating system is an obligatory reality, but that is not why you can force someone to stay still or look for elaborate solutions just to please Some.

Computers are not what they were and technology always evolves, and although it may start as something cosmetic, later it can give way to much more and if not. In any case, it is not a decision that corresponds to us to make, at least not, sitting from the comfort of our homes...

A cordial greeting to everyone.

**Comment by zahra21** [comment]
Hello again.
may i ask that notepad 7.9.4 which is released recently, can supports windows xp?
because i wished that notepad uses newer scintilla version and if i understood correctly, from now, it can open large files even with the size of more than one gb!
i really wish that i can use such version in my favorite operating system.

**Comment by sasumner** [comment]
>may i ask that notepad 7.9.4 which is released recently, can supports windows xp?

Sorry, no.  Final answer.

**Comment by donho** [comment]
@Terwilf 
Thank you for your words. They are very fair.

@KOLANICH 
I do understand your disappointment - as an user I would be upset if I were still under XP.

But my role is not a simple user, so here are some viewpoints from a project maintainer:

Is new-style dialog (or rather "modern" style) an essential functionality? It could be a debate, but the answer clearly for me is YES. It's not only about the look & feel, it's about the ergonomic issues and bunch of functionalities on which @mere-human did a lot of coding effort - the modern style dialog does matter.

My job is not to make everyone happy (whatever we do, there always are people who are not happy anyway), but execute the right tasks (with my vision) for the project - here the right task to me is scarify an old OS which had been abandoned by MS since 2014 (no more security updates or technical support from MS) to make the maintenance of Notepad++ project as easy as possible. It's a technical reason without doubt.

I apology to the XP users for the fact that Notepad++ continues without XP - you can always use the old versions under XP, and benefit the modern functionality under Windows 10 with the latest versions of Notepad++.

**Comment by KOLANICH** [comment]
>here the right task to me is scarify an old OS which had been abandoned by MS since 2014 (no more security updates or technical support from MS) to make the maintenance of Notepad++ project as easy as possible. It's a technical reason with no doubt.

No, it's not. A technical reason is when it is not technically possible. When it is possible, but someone just doesn't want, it is not a technical reason. Calling quasi-technical excuses doesn't make the reason technical, since the excuses have nothing common with the possibility of implementing this feature:

>old OS which had been abandoned by MS since 2014

How it is related to the possibility to detect the presence of the needed API in runtime?

>or technical support from MS

Do contributors of NPP really need technical support that is not available to program NPP properly? Under support I don't mean here API docs still available in the Internet, I mean what is usually meant under a support for Windows that was not available. (BTW, I once used such a support, W 7 has stopped to use a NIC, but Linux used it fine. I have phoned MS, but they haven't managed to help me. The issue has strangely resolved itself after a half of year of using Linux as a main OS: I had to run Windows and noticed that NIC icon in the tray had no ❌).

>make the maintenance of Notepad++ project as easy as possible.

Is not a technical reason. It is basically an another wording of "we don't want", and "we don't want" is not a technical reason. You have full rights not to do what you don't want to do, like keeping workaround for things noone of you needs personally. You have also a right to call it a technical reason and say any other nonsense - free speech must be an universal right. But calling the reason "we don't want to do that" technical is still disrespect to everyone reading the changelog. Please state clearly in the changelog that `<the set of devs who have enough powet to make such decisions for the whole project, and I guess @sasumner is within this set>` just don't want. It is not a very comfortable thing to say, but it is at least the truth.

**Comment by sasumner** [comment]
@donho Some people, such as KOLANICH just don't "get it".  
Ban him from the project?  :-)

**Comment by KOLANICH** [comment]
Yeah, ban me from the project, @donho. Or @sasumner would stop contributing here and the project would stagnate.

**Comment by sasumner** [comment]
>Or @sasumner would stop contributing here and the project would stagnate.

It's true.  Thank you for the compliment!  :-)

**Comment by KOLANICH** [comment]
TANSTAAFL, you know. Ones who pay (in any form) for any project decide whom to ban from there and whom not to on their own discretion.

**Comment by KOLANICH** [comment]
>Don has explained the reasons for the actions several times.

I absolutely understand the reasons not to support XP and respect your decision to drop it. Anyway, it is you bear the expenses on coding npp, not me. (If someone finds that bringing XP support back overweight the expenses of compiling and maintaining a fork, there would be one. The invisible hand.) But the reasons are not technical ("it is impossible"), but political ("we don't want") and I think it should clearly stated as such. I don't request it, but I ask you to change the text in the changelog not to create a false impression that it was dropped because it was impossible and you had no another choice, rather it was dropped because of the choices made and that there are people who are responsible for the choices made. If one makes such choices, it would be more fair to attribute the choices to himself and say "Because I have decided so. Deal with it." rather than to "circumstances", so other people would have the correct impression "it is not the circumstances and force majeure, it is choices, I can make an own fork and make own choices there".

>To continue the discussion in the manner in which you are doing it is disrespectful.

I don't think that just saying truth in a polite way is disrespectful.

**Comment by sasumner** [comment]
@KOLANICH 

Where you started by saying:

>I absolutely understand the reasons not to support XP and respect your decision to drop it...

Thank you for a nice statement. It clarifies your position.
My previous interpretation of your position was that you were campaigning to "bring XP support back" after a very-final decision.  I see now that you merely want a disavowal of a technical barrier to continuing support.

Since the reason has already been published in the change log, I don't know that a retraction and restatement of it is reasonable/possible.  If you have a suggestion for how this might be done, perhaps @donho would consider doing it.

**Comment by donho** [comment]
> Ban him from the project? :-)

No. It's not reasonable. I know you're kidding but I want to still make one point clear: 
People would be banned if they insult, spam, off-topic or do things to prevent the forum from working. Otherwise they can do any complain here freely.

BTW, thanks to @KOLANICH (and other security people) the release has been signed by GPG. Yet another reason to not ban him :)

Regarding "political reason" that @KOLANICH mentioned, as the BDFL in this project, every feature I decided to integrate is "political" according such viewpoint. Though I don't agree with such point but I will say no more about it - it'll lead meaningless & endless argue.

**Comment by zahra21** [comment]
while i wish that notepad++ continues supporting xp for more versions or forever at best!
but i am glad that notepad++ has a good and friendly community that dont insult, blame and attack lovers, users and supporters of windows xp.
i believe that you did not want to stop supporting xp. am i right?
the new feature for project panel caused notepad++ stop running under xp
and i am very glad that you did not remove supporting xp by intention!
thanks again for your great program and God bless you!

**Comment by zahra21** [comment]
mere-human,
can you please send me email?
i have a request that is not related to supporting windows xp, and i decided to send it for you directly and i wish you accept it!

**Comment by sasumner** [comment]
>mere-human,
can you please send me email?
i have a request that is not related to supporting windows xp, and i decided to send it for you directly and i wish you accept it!

This seems like an unusual request.
Of course mere-human is free to do as he likes, but...

If you have an idea for a Notepad++ feature, the proper procedure is to create an issue so that it can be discussed properly, not in some one-on-one email basis.

**Comment by zahra21** [comment]
i forgot to ask,
do developers can read my email address?
unfortunately copy link address is not available in github anymore, or at least, i could not found this item.
my email address:
nasrinkhaksar3@gmail.com
i need just one small change in notepad installer especially in 7.9.2 that i should use forever because i love xp,
and also i have some questions
and i wish to be in contact with notepad++ developers directly!

**Comment by sasumner** [comment]
>i need just one small change in notepad installer especially in 7.9.2 that i should use forever because i love xp

So that would be a feature request to be made via the issue system, but it would be declined because it is for XP.
So...don't bother.

>and also i have some questions

Questions go [HERE](https://community.notepad-plus-plus.org/)

>and i wish to be in contact with notepad++ developers directly!

It is good to want things! :-)

**Comment by zahra21** [comment]
i make issues on github for my request.
because my request is very simple to fulfill.
i want just small change in notepad++ installer, i wish the old layout of notepad websites which includes informations about all versions of notepad++ and i have some questions about using notepad++ with my screen reader.
since i think that members of notepad++ community dont use screen reader, in that case, how can they help me and answer my questions?
in the last years, i tried to subscribe in notepad++ community, but i could not and dont remember its reason now.

**Comment by mere-human** [comment]
@zahra21 
1. You can create a new issue for your request in this repository on GitHub. 
The request will be considered. 
But it is not guaranteed that the feature will be accepted by the maintainer.

2. My email is publicly visible on my GitHub page. 
Everyone can write a message. 
But I'm just a developer who contributes like any other. 
The final decision is still on Don HO who is the maintainer of this project.

3. I'd recommend to take a look at the Community Forum. 
It has a lot of active and friendly users. 
They can provide some useful tips. Or perhaps they had a similar problem and already found a solution.

**Comment by nikunjbhatt** [comment]
Most probably because of this change of using new style file/folder browser dialog, some features are missing in this dialog in Linux/Wine. (See the attached first screenshot.) It has no button to go to parent folder (or there is no item like ".." (double dots) in file/folder list to go to parent folder). (User needs to type double dots in "File Name" textbox and press enter key to go to the parent folder.)

One more problem is, it is always adding file extension of the file type selected in the drop down box of the "File Types" (located besides "File Name" textbox), even if the file name is enclosed in double quotes, except when "All types" is selected in the "File Types" dropdown. So, if user enters file name "abc.txt" (with or without double quotes) in Save dialog box, and "PHP Hypertext Preprocessor file" file type is selected in the dropdown then the file is being saved with name "abc.txt.php".

Are these problem directly related to Notepad++ or problem in Wine?

Here is my computer configuration:
Operating System: Fedora 33
KDE Plasma Version: 5.20.5
KDE Frameworks Version: 5.79.0
Qt Version: 5.15.2
Kernel Version: 5.10.21-200.fc33.x86_64
OS Type: 64-bit

Notepad++ v7.9.3   (64-bit)
Build time : Feb 14 2021 - 03:15:58
Path : C:\Program Files\Notepad++\notepad++.exe
Admin mode : ON
Local Conf mode : OFF
OS Name : Microsoft Windows 7 (64-bit) 
OS Build : 7601.0
Current ANSI codepage : 1252
WINE : 6.3
Plugins : BetterMultiSelection.dll ComparePlugin.dll DSpellCheck.dll Explorer.dll mimeTools.dll NppConverter.dll NppExport.dll zoomdisabler_x64.dll 

(Second screenshot is showing Open dialog from Wine's Notepad application.)
Open_017
Open -2-‎_020

**Comment by sasumner** [comment]
I'm not seeing Linux/WINE here: https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/SUPPORTED_SYSTEM.md
Should it be added?

**Comment by mere-human** [comment]
> It has no button to go to parent folder (or there is no item like ".." (double dots) in file/folder list to go to parent folder). (User needs to type double dots in "File Name" textbox and press enter key to go to the parent folder.)

I have the button to go to the parent folder (to the right of ← → buttons):
image

> So, if user enters file name "abc.txt" (with or without double quotes) in Save dialog box, and "PHP Hypertext Preprocessor file" file type is selected in the dropdown then the file is being saved with name "abc.txt.php".

I can't reproduce this problem. When I enter a name with some extension e.g. `new.txt` then select a different type in the dropdown, the extension is substituted to the selected one.



> Are these problem directly related to Notepad++ or problem in Wine?

It is likely that the problems you described are specific to Wine.

Also, please refrain from describing not related problems in the existing PR. 
It is better to create a separate GitHub issue, describe the problem, provide all info and keep the discussion there.

**Comment by mere-human** [comment]
Probably, a bug should be reported into Wine repository.

**Comment by nikunjbhatt** [comment]
> I have the button to go to the parent folder (to the right of ← → buttons):

This button is also appearing in the second screenshot I attached.

> Also, please refrain from describing not related problems in the existing PR (especially, closed one).

This does seems related, at least to me.

> It is better to create a separate GitHub issue, describe the problem, provide all info and keep the discussion there.
> It would be easier to address the issue in a related PR and then close the related issue once the PR is submitted.

First I wanted to confirm that whether these problems are because of Wine or Notepad++. If these are not fault of Notepad++ then creating a new issue is of no use; sasumner and donho would scold me. ;-)

**Comment by mere-human** [comment]
> > > It has no button to go to parent folder (or there is no item like ".." (double dots) in file/folder list to go to parent folder)
> >
> > I have the button to go to the parent folder (to the right of ← → buttons):
> 
> This button is also appearing in the second screenshot I attached.

The button ↑ allows you to navigate to the parent folder.
It was present in 7.9.1 (both in new style dialog and old style dialog):
image
The same button is present in 7.9.5 dialog:
image

What you have on the 1st screenshot is probably the poorly implemented feature in Wine. 
I think, their implementation of [Common Item Dialog](https://docs.microsoft.com/en-us/windows/win32/shell/common-file-dialog) is incomplete.

**Pull request closed without merging**
